File: Java/Algorithms/Moore_Algorithm.java
/*The Boyer–Moore majority vote algorithm is an algorithm for finding the majority of a sequence of elements using linear time and constant space. */

import java.util.*;

public class MajorityVote {

	// naive approach
	public static void check(int arr[]) {
		int count = 0;
		for (int i = 0; i < arr.length - 1; i++) {
			count = 1;
			for (int j = i + 1; j < arr.length; j++) {
				// Counting the number by matching it with other numbers.
				if (arr[j] == arr[i]) {
					count++;
				}
			}
			if (count > arr.length / 2) {
				System.out.println("Number Occuring More than 'n/2'=" + arr[i]);
				System.exit(0);
			}
		}
		System.out.print("NO majority Elements ");
	}

	// using The Boyer–Moore majority vote algorithm
	public static int majority(int arr[], int arrsize) {

		int index = 0, count = 1;
		int i;
		// finding num to check majority
		for (i = 1; i < arrsize; i++) {
			if (arr[index] == arr[i])
				count++;
			else
				count--;
			if (count == 0) {
				index = i;
				count = 1;
			}
		}
		int num = arr[index];
		for (i = 0; i < arrsize; i++)
			if (arr[i] == num)
				count++;
		// checking for majority
		if (count > arrsize / 2)
			return num;
		else
			return -1;
	}

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		System.out.println("Enter Number of elements (Eg. 6) : ");
		int element_count = input.nextInt();
		int[] arr = new int[element_count];
		System.out.println("Enter " + element_count + " Numbers:");
		for (int i = 0; i < element_count; i++) {
			// Giving values to array arr
			arr[i] = input.nextInt();
		}
		// Passing array to function
		check(arr);
		// using boyer-moore algorithm
		int result = majority(arr, element_count);
		if (result != -1)
			System.out.println("Majority of a sequence is " + result);

		else
			System.out.println("There is no majority \n");

	}

}

/*
 * Naive method: 
 * Enter Number of elements : 6 
 * Enter 6 Numbers: 5 5 5 5 4 2
 * Number Occuring More than 'n/2'=5 
 * Time Complexity : o(n^2) , Space Complexity : o(1).
 * 
 * The Boyer–Moore majority vote algorithm: 
 * Enter no of elements:5 
 * Enter elements: 1 1 1 1 2 
 * Majority of arr sequence is 1
 * 
 * Time complexity : O(n) Space complexity: O(1)
 */

File: Python/other/Moore_Algorithm.py
#The Boyer–Moore majority vote algorithm is an algorithm for finding the majority of a sequence of elements using linear time and constant space.
def counting_fun(array):
    count_temp = 0                       # stores occurence count
    for i in range(0, len(array)):
        count_temp = array.count(array[i])   # Counts the occurence
        if count_temp > number/2:
            print(f"Number occuring more than n/2 is :{array[i]}")
            count_temp = "True"
            break
    if(count_temp != "True"):
        print("No Element occuring more than n/2")
number = int(input("Enter Number of values you will enter(eg.6): "))
i = 0
list_1 = []                         # List Declaration
print(f"Enter {number} values: ")
for i in range(0, number):
    elements = int(input())
    list_1.append(elements)         # adding element to the list

counting_fun(list_1)

# Output:
# Enter 6 values:
# 1
# 1
# 1
# 1
# 4
# 2
# Number occuring more than n/2 is: 1
# Time Complexity : o(n) , Space Complexity : o(1)
